rfc/rfc000-package-server.md

```diff
@@ -1,19 +1,91 @@
+# Package Server
+
 ---
+
 RFC: 000
-Author: Peter Bourgon <peter@bourgon.org>
+
+Author: Matt Heath <matt@mattheath.com>
+
 Status: Draft
----
 
-# package server
+---
 
 ## Motivation
 
-http://peter.bourgon.org/go-kit/#package-server
+[http://peter.bourgon.org/go-kit/#package-server](http://peter.bourgon.org/go-kit/#package-server)
+
+Package server is probably the biggest and most important component of the toolkit. Ideally, we should be able to write our services as implementations of normal, nominal Go interfaces, and delegate integration with the environment to the server package. The package should encode and enforce conventions for server-side concerns, like health checks, system-wide request tracing, connection management, backpressure and throttling, and so on. For each of those topics, it should provide interfaces for different, pluggable strategies. It should integrate with service discovery, and work equally well over multiple transports. Considerable prior art exists in the form of Finagle, Karyon (Netflixs application service library), and likely many more.
 
 ## Scope
 
-To be defined.
+### Endpoints
+
+*   An endpoint is defined as a handler interface which receives a request and returns a response or an error to the client.
```

Comment:
@peterbourgon as you know via twitter, scope creep for things like this is terrifying!

I don't really care about supporting other message-passing paradigms from day one. The trouble is that it can be really difficult to retrofit them onto a scheme that assumes RPC for everything, mainly since RPC looks so much like a function call and thus implies all sorts of API decisions that simply don't work for other paradigms. What excites me about Go in this instance is that the channel primitive is such a good fit for a few of the other big/important paradigms.

Of course this is your baby, so do what you will. If it was me, though, I would at least support a mode where the server can return N values as a "streaming" response. This is critically important for microservices that need to return objects that are as large or larger than their own memory footprint. You *need* to break the objects up into pieces, and with streaming responses you are saved a lot of trouble and complexity. (I.e., stick with only one request object to keep the interface closer to RPC)

Another possibility would be to make sure that all of the auth stuff is well-encapsulated in the transport layer and tell people who need streaming (etc) to build it themselves on top of that encapsulated transport layer.